components/icu/README.md

```diff
@@ -35,20 +35,22 @@ for a small set of locales.
 ## Example
 
 ```rust
-use icu::datetime::{mock::parse_gregorian_from_str, options::length, TypedDateTimeFormatter};
+use icu::calendar::DateTime;
+use icu::datetime::{options::length, DateTimeFormatter};
 use icu::locid::locale;
 
 let provider = icu_testdata::get_provider();
 
 let options =
     length::Bag::from_date_time_style(length::Date::Long, length::Time::Medium).into();
 
-let dtf = TypedDateTimeFormatter::try_new_with_buffer_provider(&provider, &locale!("es").into(), options)
+let dtf = DateTimeFormatter::try_new_with_buffer_provider(&provider, &locale!("es").into(), options)
     .expect("Failed to create TypedDateTimeFormatter instance.");
 
-let date = parse_gregorian_from_str("2020-09-12T12:35:00").expect("Failed to parse date.");
+let date = DateTime::new_gregorian_datetime(2020, 9, 12, 12, 35, 0).expect("Failed to parse date.");
+let date = date.to_any();
```

Comment:
Issue (here and elsewhere): this only works because locale "es" defaults to gregorian calendar, right? Please use `new_iso_datetime` instead.

We should be using `new_iso_datetime` when we want the date to be unopinionated about its calendar system and automatically pick up the one from the formatter. It is harmful to the ecosystem if people willy-nilly create gregorian-tagged AnyCalendar datetimes, because they will inevitably pass them to a locale that uses a non-gregorian calendar and then get an error.

Optional extension: I'm also now wondering whether we should make `.to_any()` so easy to use, because it seems likely to be used in situations like this where it is wrong. _Most_ AnyCalendar datetimes should be ISO, unless arithmetic is being performed. Maybe the short function name `.to_any()` should be `.to_any_iso()`, and the one that retails the calendar should look more scary like `.to_any_retain_calendar()` or something.

We could also consider just always auto-converting in ICU4X DateTimeFormat instead of returning an error. The difference with ECMAScript Temporal and the reason it is stricter in this area is that it is intended to be hard to _accidentally_ create calendar-specific datetimes.

Reply:
> Most AnyCalendar datetimes should be ISO

I disagree: The way I see it is that this crate is not a formatting crate: it is more fundamental than that, and in that context I find to_any to be a very useful method to have. I think "most datetimes are iso" makes some sense in the formatting sense (though I'm not completely convinced of that either), but not enough to impact this crate.


Having DTF always convert is definitely something I'm open to. I'm a bit worried about codesize impact but I think DTF _already_ needs to carry around a ton of conversion code. (And conversion code isn't actually that expensive)

Reply:
(I do agree that all DTF examples should move to iso though. I made the change)

Reply:
> Most AnyCalendar datetimes should be ISO

My full sentence was

> _Most_ AnyCalendar datetimes should be ISO, unless arithmetic is being performed.

with the second part of the sentence being very important. And it sounds like we agree: if you are using these purely as input to DateTimeFormatter, then you should most likely choose ISO. If you are doing something more interesting with them, then you may want a localized calendar.

My concern is that since these types are the primary input types for DateTimeFormatter, that is going to be the primary use case for these types, and we should avoid APIs that let people do the wrong thing accidentally.

Reply:
I think amongst ICU4X's _own_ uses they're used for formatting but I think as a library in isolation they might be used for many things. Certainly my expectation on seeing a calendars library would be that such transforms would be easy.